File: backend/internal/model/pipeline.go
package model

import "time"

type Pipeline struct {
	ID               uint            `gorm:"primaryKey" json:"id"`
	PipelineNo       string          `gorm:"uniqueIndex;size:32;not null" json:"pipelineNo"`
	TaskID           uint            `json:"taskId"`
	ApplicationID    uint            `json:"applicationId"`
	AppName          string          `gorm:"size:128;not null" json:"appName"`
	AppType          string          `gorm:"size:20" json:"appType"`
	GitRepo          string          `gorm:"size:512" json:"gitRepo"`
	GitBranch        string          `gorm:"size:128" json:"gitBranch"`
	GitCommitHash    string          `gorm:"size:64" json:"gitCommitHash"`
	GitCommitAuthor  string          `gorm:"size:64" json:"gitCommitAuthor"`
	GitCommitMessage string          `gorm:"type:text" json:"gitCommitMessage"`
	Status           string          `gorm:"size:20;not null;default:PENDING" json:"status"` // PENDING/RUNNING/IMAGE_READY/DEPLOYING/SUCCESS/FAILED/CANCELED
	TriggerUser      string          `gorm:"size:64;not null" json:"triggerUser"`
	BuilderImage     string          `gorm:"size:256" json:"builderImage"`
	ArtifactName     string          `gorm:"size:128" json:"artifactName"`
	UploadPath       string          `gorm:"size:512" json:"uploadPath"`
	UploadTargets    string          `gorm:"size:512" json:"uploadTargets"`
	BatchID          string          `gorm:"size:64;index" json:"batchId"`
	DeployMode       string          `gorm:"size:32" json:"deployMode"` // upload_only / upload_and_deploy
	ImageName        string          `gorm:"size:256" json:"imageName"` // Built Docker image name (e.g. registry/app:tag)
	K8sNamespace     string          `gorm:"size:64" json:"k8sNamespace"` // K8s namespace
	StartTime        *time.Time      `json:"startTime"`
	EndTime          *time.Time      `json:"endTime"`
	DurationSeconds  *int            `json:"durationSeconds"`
	ErrorStage       string          `gorm:"size:64" json:"errorStage"`
	ErrorMessage     string          `gorm:"type:text" json:"errorMessage"`
	Stages           []PipelineStage `gorm:"foreignKey:PipelineID" json:"stages,omitempty"`
	CreatedAt        time.Time       `json:"createdAt"`
}

File: frontend/src/api/pipeline.ts
import request, { type PageResult } from './request'

export interface PipelineStage {
  id: number
  pipelineId: number
  stageCode: string
  stageName: string
  stageOrder: number
  status: string
  command?: string
  startTime?: string | null
  endTime?: string | null
  durationSeconds?: number | null
  exitCode?: number | null
  errorMessage?: string
}

export interface Pipeline {
  id: number
  pipelineNo: string
  taskId: number
  applicationId: number
  appName: string
  appType: string
  gitRepo: string
  gitBranch: string
  gitCommitHash?: string
  gitCommitAuthor?: string
  gitCommitMessage?: string
  status: string
  triggerUser: string
  builderImage?: string
  artifactName?: string
  uploadPath?: string
  uploadTargets?: string
  batchId?: string
  deployMode?: string
  startTime?: string | null
  endTime?: string | null
  durationSeconds?: number | null
  errorStage?: string
  errorMessage?: string
  stages?: PipelineStage[]
  createdAt?: string
}

export interface StageLog {
  id: number
  pipelineId: number
  stageId: number
  lineNumber: number
  content: string
  stream: string
  timestamp: string
}

export function listPipelines(params?: { page?: number; pageSize?: number; app?: string; status?: string }) {
  return request.get<any, PageResult<Pipeline>>('/pipelines', { params })
}

export function getPipeline(id: number) {
  return request.get<any, Pipeline>(`/pipelines/${id}`)
}

export function cancelPipeline(id: number) {
  return request.post<any, null>(`/pipelines/${id}/cancel`)
}

export function getStageLogs(pipelineId: number, stageId: number) {
  return request.get<any, StageLog[]>(`/pipelines/${pipelineId}/stages/${stageId}/logs`)
}

export function getBuildQueue() {
  return request.get<any, { running: Pipeline[]; pending: Pipeline[] }>('/build-queue')
}

export function cancelQueued(id: number) {
  return request.delete<any, null>(`/build-queue/${id}`)
}

export function deployPipeline(id: number) {
  return request.post<any, null>(`/pipelines/${id}/deploy`)
}

export function streamLogsUrl(pipelineId: number): string {
  // SSE EventSource doesn't send Authorization header by default
  // Backend needs to accept token via query param for SSE
  return `/api/pipelines/${pipelineId}/logs/stream`
}

// ---------- Dashboard ----------
export interface DashboardStats {
  totalApps: number
  todayBuilds: number
  successCount: number
  failedCount: number
}

export const getDashboardStats = () => request.get<any, DashboardStats>('/dashboard/stats')
export const getRecentBuilds = () => request.get<any, Pipeline[]>('/dashboard/recent-builds')

// ---------- Artifact ----------
export interface Artifact {
  id: number
  pipelineId: number
  pipelineNo: string
  appName: string
  artifactName: string
  gitBranch: string
  gitCommitHash?: string
  uploadPath: string
  uploadTargets: string
  sourceType?: string
  sourcePath?: string
  storagePath?: string
  sha256?: string
  batchId?: string
  isLatest?: boolean
  fileSizeBytes?: number
  durationSeconds: number
  createdAt: string
}

export function listArtifacts(params?: { page?: number; pageSize?: number; search?: string }) {
  return request.get<any, PageResult<Artifact>>('/artifacts', { params })
}

// ---------- Remote Tools ----------
export function remoteSyncUrl() {
  return '/api/remote/sync'
}

export function remotePackageUrl() {
  return '/api/remote/package'
}
